File: frontend/lib/model/routine.dart
import 'package:flutter/material.dart';
import 'package:frontend/model/abstracts.dart';
import 'package:json_annotation/json_annotation.dart';

part 'routine.g.dart';

@JsonEnum()
enum RoutineType { other, exercises, meal }

@JsonSerializable(explicitToJson: true)
class Routines implements ListAbstract<Routine> {
  @override
  @JsonKey(includeFromJson: false, includeToJson: false)
  IconData icon = Icons.inbox_outlined;

  @override
  String collection = '';

  @override
  String id = '';

  @override
  List<Routine> items = const <Routine>[];

  String? wakeupTime;

  Routines();

  factory Routines.fromJson(Map<String, dynamic> json) => _$RoutinesFromJson(json);
  Map<String, dynamic> toJson() => _$RoutinesToJson(this);
  
  @override
  Routine createNewItem() {
    return Routine()..name = '';
  }

}

@JsonSerializable()
class Routine extends CardAbstract {
  String name = '';
  
  @JsonKey(defaultValue: false, includeFromJson: false, includeToJson: false)
  bool isExpanded = false;
  
  @JsonKey(defaultValue: false)
  bool isCompleted = false;
  
  @override
  @JsonKey(includeFromJson: false, includeToJson: false)
  bool canAddItems = false;
  
  int? durationMin;
  
  @JsonKey(defaultValue: RoutineType.other)
  RoutineType routineType = RoutineType.other;
  
  dynamic ref;

  Routine();

  factory Routine.fromJson(Map<String, dynamic> json) => _$RoutineFromJson(json);
  Map<String, dynamic> toJson() => _$RoutineToJson(this);



  @override
  List<FieldDescriptor> getEditableFields() {
    return [
      FieldDescriptor('name', 'Name', () => name, (v) => name = v, String),
      FieldDescriptor('durationMin', 'Duration (min)', () => durationMin, (v) => durationMin = v, int),
      FieldDescriptor('routineType', 'Routine Type', () => routineType.name, (v) => routineType = RoutineType.values.byName(v), String),
      FieldDescriptor('isCompleted', 'Completed', () => isCompleted, (v) => isCompleted = v, bool),
    ];
  }

}

File: react-native-expo/models/Routine.ts
import { CardAbstract } from './abstract/CardAbstract';
import { FieldDescriptor } from './utils';

/**
 * Routine type enumeration
 * Equivalent to Dart's RoutineType enum
 */
export enum RoutineType {
  OTHER = 'other',
  EXERCISES = 'exercises',
  MEALS = 'meals'
}

/**
 * Routine model for routine cards
 * Equivalent to Dart's Routine class
 */
export class Routine extends CardAbstract {
  isExpanded: boolean = false;
  canAddItems: boolean = false;
  durationMin?: number;
  routineType: RoutineType = RoutineType.OTHER;
  ref?: any; // Reference to other objects (exercises, meals, etc.)

  constructor(data?: Partial<Routine>) {
    super();
    if (data) {
      // Handle routineType enum conversion
      if (data.routineType && typeof data.routineType === 'string') {
        this.routineType = data.routineType as RoutineType;
        delete data.routineType;
      }
      Object.assign(this, data);
    }
  }

  /**
   * Get editable fields for dynamic form generation
   */
  getEditableFields(): FieldDescriptor[] {
    return [
      new FieldDescriptor(
        'name',
        'Name',
        () => this.name,
        (v: string) => this.name = v,
        'string'
      ),
      new FieldDescriptor(
        'durationMin',
        'Duration (min)',
        () => this.durationMin,
        (v: number) => this.durationMin = v,
        'number'
      ),
      new FieldDescriptor(
        'routineType',
        'Routine Type',
        () => this.routineType,
        (v: string) => this.routineType = v as RoutineType,
        'string'
      ),
      new FieldDescriptor(
        'isCompleted',
        'Completed',
        () => this.isCompleted,
        (v: boolean) => this.isCompleted = v,
        'boolean'
      ),
    ];
  }

  /**
   * Get computed tags for display
   */
  getTags(): string[] {
    const tags: string[] = [];
    if (this.durationMin) {
      tags.push(`${this.durationMin} min`);
    }
    if (this.routineType !== RoutineType.OTHER) {
      tags.push(this.routineType);
    }
    return tags;
  }

  /**
   * Routines don't typically have sub-items
   */
  createNewItem(): null {
    return null;
  }

  /**
   * Check if this routine is a reference to another collection
   */
  isReference(): boolean {
    return this.routineType !== RoutineType.OTHER;
  }

  /**
   * Get display name with duration if available
   */
  getDisplayName(): string {
    if (this.durationMin && this.durationMin > 0) {
      return `${this.name} (${this.durationMin} min)`;
    }
    return this.name;
  }
}
